bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java

```diff
@@ -0,0 +1,529 @@
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.StateOption;
+import org.openhab.binding.intesis.internal.IntesisConfiguration;
+import org.openhab.binding.intesis.internal.IntesisDynamicStateDescriptionProvider;
+import org.openhab.binding.intesis.internal.IntesisHomeModeEnum;
+import org.openhab.binding.intesis.internal.api.IntesisHomeHttpApi;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Data;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Datapoints;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Descr;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Dp;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Dpval;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Id;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Info;
+import org.openhab.binding.intesis.internal.gson.IntesisHomeJSonDTO.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link IntesisHomeHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ */
+@NonNullByDefault
+public class IntesisHomeHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(IntesisHomeHandler.class);
+    private final IntesisHomeHttpApi api;
+    private IntesisConfiguration config = new IntesisConfiguration();
+
+    private final Map<String, String> properties = new HashMap<>();
+
+    private IntesisDynamicStateDescriptionProvider intesisStateDescriptionProvider;
+
+    private @Nullable ScheduledFuture<?> refreshJob;
+
+    private String ipAddress = "";
+    private String password = "";
+    private String sessionId = "";
+
+    final Gson gson = new Gson();
+
+    public IntesisHomeHandler(final Thing thing, final HttpClient httpClient,
+            IntesisDynamicStateDescriptionProvider intesisStateDescriptionProvider) {
+        super(thing);
+        this.api = new IntesisHomeHttpApi(config, httpClient);
+        this.intesisStateDescriptionProvider = intesisStateDescriptionProvider;
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        final IntesisConfiguration config = getConfigAs(IntesisConfiguration.class);
+        ipAddress = config.ipAddress;
+        password = config.password;
+        if (ipAddress.isEmpty() || password.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+        }
+
+        // start background initialization:
+        scheduler.schedule(() -> {
+            populateProperties();
+            getDataPoints();
```

Comment:
What if either of these methods fail? Would the thing properties never get populated? Perhaps you could make it retry on failure.

Reply:
I don't see a reason for a retry and did not see such a thing in other bindings. If either of these fail, the thing status is set to offline, as the is some kind of communication error.